ray-operator/controllers/ray/batchscheduler/scheduler-plugins/scheduler_plugins.go

```diff
@@ -0,0 +1,135 @@
+package schedulerplugins
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	schedulerinterface "github.com/ray-project/kuberay/ray-operator/controllers/ray/batchscheduler/interface"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+)
+
+const (
+	SchedulerName                 string = "kube-scheduler"
+	KubeSchedulerPodGroupLabelKey string = "scheduling.x-k8s.io/pod-group"
+)
+
+type KubeScheduler struct {
+	cli client.Client
+}
+
+type KubeSchedulerFactory struct{}
+
+func GetPluginName() string {
+	return SchedulerName
+}
+
+func (k *KubeScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (k *KubeScheduler) DoBatchSchedulingOnSubmission(ctx context.Context, rc *rayv1.RayCluster) error {
+	if !k.isGangSchedulingEnabled(rc) {
+		return nil
+	}
+	replica := int32(1)
+	for _, workerGroup := range rc.Spec.WorkerGroupSpecs {
+		if workerGroup.Replicas == nil {
+			continue
+		}
+		replica += *workerGroup.Replicas
+	}
+
+	podGroup := &v1alpha1.PodGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       rc.Name,
+					UID:        rc.UID,
+					APIVersion: rc.APIVersion,
+					Kind:       rc.Kind,
+				},
+			},
+		},
+		Spec: v1alpha1.PodGroupSpec{
+			MinMember:    replica,
+			MinResources: utils.CalculateMinResources(rc),
+		},
+	}
+
+	err := k.cli.Create(ctx, podGroup)
+	if errors.IsAlreadyExists(err) {
+		return nil
+	}
+	return err
+}
+
+// AddMetadataToPod adds essential labels and annotations to the Ray pods
+// the scheduler needs these labels and annotations in order to do the scheduling properly
+func (k *KubeScheduler) AddMetadataToPod(_ context.Context, app *rayv1.RayCluster, groupName string, pod *corev1.Pod) {
```

Comment:
s/app/cluster

Reply:
Could you please explain the comment? Thank you!